### PR TITLE
[URP-220] Add URP backbuffer HDR format options

### PR DESCRIPTION
**Pull request by eh-unity**
### Checklist for PR maker
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Add FP16 rendertarget/backbuffer format for URP to fix blue/yellow banding cases at the cost of perf.

---
### Testing status
Describe what manual/automated tests were performed for this PR

Added a test case:
Fail:
161_FP16_BackBuffer
Pass:
161_FP16_BackBuffer
---
### UI

URPAsset:
image
image
image
image
image

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 1]
```diff
@@ -46,6 +46,13 @@ public enum LightCookieFormat
         ColorHDR,
     }
 
+    // TODO: should this be like light cookie format, an index to a  list of formats that are tested in order for support.
+    public enum HDRFormat
+    {
+        Default = GraphicsFormat.B10G11R11_UFloatPack32,
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:52]: Explicit formats as enum values? Or just ints and a function to get the formats (which are of course enum ints already).

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs [hunk 2]
```diff
@@ -98,16 +98,42 @@ public class PostProcessPass : ScriptableRenderPass
...
+        {
+            m_BlitMaterial = postProcessParams.blitMaterial;
+
+            // Texture format pre-lookup
+            const FormatUsage usage = FormatUsage.Linear | FormatUsage.Render;
+            if (SystemInfo.IsFormatSupported(postProcessParams.requestHDRFormat, usage))
+            {
+                // TODO: we could get render target format from
+                // static `UniversalRenderPipeline.asset.hdrFormat´ but that would means a static tight dependency to the pipeline asset.
+
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:111]: Feedback: Do we want to rely asset being always available? and add dependencies? I'd say no.

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs [hunk 3]
```diff
@@ -30,28 +45,40 @@ internal struct PostProcessPasses : IDisposable
...
+            m_BlitMaterial = null;
 
             m_AfterPostProcessColor = null;
             m_ColorGradingLut = null;
-
             m_RendererPostProcessData = rendererPostProcessData;
-            m_BlitMaterial = blitMaterial;
+        }
 
+        public PostProcessPasses(PostProcessData rendererPostProcessData, ref PostProcessParams postProcessParams) : this(rendererPostProcessData)
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs:61]: Refactored this to have a struct interface for cleaner future param additions.

Feedback? C# is a bit clumsy for the backcompat coding, better ideas?

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 4]
```diff
@@ -579,8 +581,14 @@ public ScriptableRenderer(ScriptableRendererData data)
             Clear(CameraRenderType.Base);
             m_ActiveRenderPassQueue.Clear();
 
+            // TODO: should the asset be passed in to break deps? Maybe we can then rely on the asset, an no need to duplicate state here.
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:584]: Feedback? Defensive asset programming here. Replicate data or pass asset forward here? Or just rely that asset always exists. (that will add deps to everywhere).

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs [hunk 5]
```diff
@@ -557,11 +557,12 @@ void SortCameras(Camera[] cameras)
...
 #endif
 
-        static GraphicsFormat MakeRenderTextureGraphicsFormat(bool isHdrEnabled, bool needsAlpha)
+        internal static GraphicsFormat MakeRenderTextureGraphicsFormat(bool isHdrEnabled, HDRFormat requestHdrFormat, bool needsAlpha)
         {
             if (isHdrEnabled)
             {
-                if (!needsAlpha && RenderingUtils.SupportsGraphicsFormat(GraphicsFormat.B10G11R11_UFloatPack32, FormatUsage.Linear | FormatUsage.Render))
+                // TODO: we need a proper format scoring system. Score formats, sort, pick first or pick first supported (if not in score).
+                if (!needsAlpha && requestHdrFormat != HDRFormat.High && RenderingUtils.SupportsGraphicsFormat(GraphicsFormat.B10G11R11_UFloatPack32, FormatUsage.Linear | FormatUsage.Render))
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:565]: Feedback? so we could just cast the HDRFormat to GraphicsFormat here or similar sitautions, since it's values are GraphicsFormat. Opinions?

Also adding formats this way doesn't really scale, we need a better system if we want more formats cleanly.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs [hunk 6]
```diff
@@ -119,6 +145,14 @@ public PostProcessPass(RenderPassEvent evt, PostProcessData data, Material blitM
                     : GraphicsFormat.R8G8B8A8_UNorm;
                 m_UseRGBM = true;
             }
+        }
+
+        private PostProcessPass(RenderPassEvent evt, PostProcessData data)
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:150]: Feedback? Struct interface backcompat consequence. Better ideas?

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs [hunk 7]
```diff
@@ -30,28 +45,40 @@ internal struct PostProcessPasses : IDisposable
...
+        {
+            m_BlitMaterial = postProcessParams.blitMaterial;
+            Recreate(rendererPostProcessData, ref postProcessParams);
+        }
 
-            Recreate(rendererPostProcessData);
+        [Obsolete]
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs:67]: Same as the other class, this one is in internal namespace so you can just remove the constructor.

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs [hunk 8]
```diff
@@ -75,12 +102,19 @@ public void Recreate(PostProcessData data)
...
                 m_ColorGradingLutPass = new ColorGradingLutPass(RenderPassEvent.BeforeRenderingPrePasses, data);
-                m_PostProcessPass = new PostProcessPass(RenderPassEvent.BeforeRenderingPostProcessing, data, m_BlitMaterial);
-                m_FinalPostProcessPass = new PostProcessPass(RenderPassEvent.AfterRenderingPostProcessing, data, m_BlitMaterial);
+                m_PostProcessPass = new PostProcessPass(RenderPassEvent.BeforeRenderingPostProcessing, data, ref ppParams);
+                m_FinalPostProcessPass = new PostProcessPass(RenderPassEvent.AfterRenderingPostProcessing, data, ref ppParams);
                 m_CurrentPostProcessData = data;
             }
         }
 
+        [Obsolete]
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs:111]: Same here as well.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 9]
```diff
@@ -499,6 +499,8 @@ internal struct RTHandleRenderTargetIdentifierCompat
         RTHandleRenderTargetIdentifierCompat m_CameraDepthTarget;
         RTHandleRenderTargetIdentifierCompat m_CameraResolveTarget;
 
+        HDRFormat m_HDRFormat;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:502]: I wonder if you could store and initialize this as part of CameraData instead.

Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:502]: I guess you actually already have this in `cameraData.cameraTargetDescriptor.graphicsFormat` as well which you initialize here? https://github.com/Unity-Technologies/Graphics/pull/6728/files?file-filters%5B%5D=.cs&show-viewed-files=true#diff-0eb17165c501725142edafebc19d4d08538deb40c78a8cb059741641db83581cR1287

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 4 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:584]: It's not replication as the settings in asset means requested settings, you need to have the runtime platform data settings anyway.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs (hunk 5 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:565]: I think casting makes sense.

**Review by phi-lira — CHANGES_REQUESTED** [review]

File: com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAsset/UniversalRenderPipelineAssetUI.Skin.cs [hunk 10]
```diff
@@ -34,6 +34,7 @@ internal static class Styles
 
             // Quality
             public static GUIContent hdrText = EditorGUIUtility.TrTextContent("HDR", "Controls the global HDR settings.");
+            public static GUIContent hdrFormatText = EditorGUIUtility.TrTextContent("HDR Format", "Controls the global HDR settings.");
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Editor/UniversalRenderPipelineAsset/UniversalRenderPipelineAssetUI.Skin.cs:37]: Ah, it looks like it's copy paste from above line. I'll fix it.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 1 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:52]: Ok. Makes sense. I'll change it to match that.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs (hunk 6 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:150]: This is the private common part, the public api part inits the rest. This is only done to keep the old method.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs [hunk 11]
```diff
@@ -98,16 +98,42 @@ public class PostProcessPass : ScriptableRenderPass
...
+            {
+                m_DefaultHDRFormat = QualitySettings.activeColorSpace == ColorSpace.Linear
+                    ? GraphicsFormat.R8G8B8A8_SRGB
+                    : GraphicsFormat.R8G8B8A8_UNorm;
+                m_UseRGBM = true;
+            }
+        }
+
+        [Obsolete]
+        public PostProcessPass(RenderPassEvent evt, PostProcessData data, Material blitMaterial) : this(evt, data)
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:130]: Oh, if I've missed the class visibility then this is good news. No need to do this.

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs [hunk 12]
```diff
@@ -2,10 +2,25 @@
...
+    public struct PostProcessParams
+    {
+        public Material blitMaterial;
+        public GraphicsFormat requestHDRFormat;
+
+        public static PostProcessParams GetDefault()
+        {
+            PostProcessParams ppParams;
+            ppParams.blitMaterial = null;
+            ppParams.requestHDRFormat = GraphicsFormat.B10G11R11_UFloatPack32;
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs:19]: I think it's based on the old code. Anyway it's just something sensible to init the struct.

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs (hunk 7 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs:67]: Yeah. I guess I've originally forgotten to look at the class visibility then.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs [hunk 13]
```diff
@@ -98,16 +98,42 @@ public class PostProcessPass : ScriptableRenderPass
...
+            const FormatUsage usage = FormatUsage.Linear | FormatUsage.Render;
+            if (SystemInfo.IsFormatSupported(postProcessParams.requestHDRFormat, usage))
+            {
+                // TODO: we could get render target format from
+                // static `UniversalRenderPipeline.asset.hdrFormat´ but that would means a static tight dependency to the pipeline asset.
+
+                m_DefaultHDRFormat = postProcessParams.requestHDRFormat;
+                m_UseRGBM = false;
+            }
+            else if (SystemInfo.IsFormatSupported(GraphicsFormat.B10G11R11_UFloatPack32, usage)) // HDR fallback
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:115]: I think this one is for post processing internal LUT (a different thing) and can be different.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs (hunk 2 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:111]: Data is the resources.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 1 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:52]: I changed the name to ColorBufferFormat. However, it explicitly defines the formats. 
In the past we've chosen to use abstract Low, High etc. to be more user friendly (less technical) and to be able to change the formats easier. I've kept that aspect. Any thoughts?

Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:52]: Actually, this is also only for HDR and we also have SDR. I don't think it makes sense share this. I've added the ColorBuffer part to the name, since it's more exact and there can be many "HDRFormats".

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 14]
```diff
@@ -169,6 +176,7 @@ public partial class UniversalRenderPipelineAsset : RenderPipelineAsset, ISerial
 
         // Quality settings
         [SerializeField] bool m_SupportsHDR = true;
+        [SerializeField] HDRFormat m_HDRFormat = HDRFormat.Default; // TODO: could just replace bool with the enum
```
Comment (eh-unity) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:179]: We could merge supports HDR to HDRFormat as false == None.
We could probably merge also SDR formats into this enum. Just have a ColorBuffer enum for SDR and HDR.
Thoughts?

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 15]
```diff
@@ -665,6 +673,12 @@ public bool supportsHDR
             set { m_SupportsHDR = value; }
         }
 
+        public HDRFormat hdrFormat
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:676]: done.

File: com.unity.render-pipelines.universal/Runtime/NativeRenderPass.cs [hunk 16]
```diff
@@ -229,7 +229,7 @@ internal void SetNativeRenderPassMRTAttachmentList(ScriptableRenderPass renderPa
                     for (int i = 0; i < validColorBuffersCount; ++i)
                     {
                         AttachmentDescriptor currentAttachmentDescriptor =
-                            new AttachmentDescriptor(pass.renderTargetFormat[i] != GraphicsFormat.None ? pass.renderTargetFormat[i] : GetDefaultGraphicsFormat(cameraData));
+                            new AttachmentDescriptor(pass.renderTargetFormat[i] != GraphicsFormat.None ? pass.renderTargetFormat[i] : UniversalRenderPipeline.MakeRenderTextureGraphicsFormat(cameraData.isHdrEnabled, m_HDRFormat, Graphics.preserveFramebufferAlpha));
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/NativeRenderPass.cs:232]: done

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs (hunk 7 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs:67]: fixed.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 9 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:502]: done.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs (hunk 6 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:150]: removed the backwards compatibility.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs [hunk 17]
```diff
@@ -19,7 +19,7 @@ namespace UnityEngine.Rendering.Universal.Internal
     /// <summary>
     /// Renders the post-processing effect stack.
     /// </summary>
-    public class PostProcessPass : ScriptableRenderPass
+    internal class PostProcessPass : ScriptableRenderPass
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:22]: you should change the namespace as well to be `UnityEngine.Rendering.Universal`. We should actually get rid of `UnityEngine.Rendering.Universal.Internal` for all passes, but I recommend to be done in another PR.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 18]
```diff
@@ -579,8 +579,11 @@ public ScriptableRenderer(ScriptableRendererData data)
             Clear(CameraRenderType.Base);
             m_ActiveRenderPassQueue.Clear();
 
+            // TODO: should the asset be passed in to break deps? Maybe we can then rely on the asset, an no need to duplicate state here.
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:582]: Should this TODO be removed?

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs [hunk 19]
```diff
@@ -139,8 +139,9 @@ public Matrix4x4 GetGPUProjectionMatrix(int viewIndex = 0)
         public float renderScale;
         internal ImageScalingMode imageScalingMode;
         internal ImageUpscalingFilter upscalingFilter;
-        public bool clearDepth;
+        public HDRColorBufferFormat hdrColorBufferFormat;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:142]: Needs public API docs.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs [hunk 20]
```diff
@@ -139,8 +139,9 @@ public Matrix4x4 GetGPUProjectionMatrix(int viewIndex = 0)
...
         public CameraType cameraType;
+        public bool clearDepth;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:144]: Needs public API docs?

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderer.cs [hunk 21]
```diff
@@ -1,4 +1,5 @@
 using System.Collections.Generic;
+using UnityEditor.VersionControl;
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderer.cs:2]: Was this added by mistake?

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs [hunk 22]
```diff
@@ -30,7 +52,12 @@ internal struct PostProcessPasses : IDisposable
 
         public bool isCreated { get => m_CurrentPostProcessData != null; }
 
-        public PostProcessPasses(PostProcessData rendererPostProcessData, Material blitMaterial)
+        /// <summary>
+        /// Creates post process passes with supplied data anda params.
+        /// </summary>
+        /// <param name="rendererPostProcessData">Post process resources.</param>
+        /// <param name="postProcessParams">Post process run-time creation parameters.</param>
+        public PostProcessPasses(PostProcessData rendererPostProcessData, ref PostProcessParams postProcessParams)
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs:60]:
```suggestion
        public PostProcessPasses(PostProcessData rendererPostProcessData, in PostProcessParams postProcessParams)
```

Imo we should use `in` which is the C# const reference. The reason we have several signatures with `ref` today even though they should be immutable is because `in` was not available in C# by the time we started URP development.

**Review by phi-lira — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderer.cs (hunk 21 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderer.cs:2]: Yes. I think Rider IDE has auto-added it for some reason and it broke all the tests.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs (hunk 20 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:144]: I just moved the bools together as minor cleanup.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs [hunk 23]
```diff
@@ -788,7 +788,7 @@ static void InitializeCameraData(Camera camera, UniversalAdditionalCameraData ad
             float renderScale = isSceneViewCamera ? 1.0f : cameraData.renderScale;
 
             cameraData.cameraTargetDescriptor = CreateRenderTextureDescriptor(camera, renderScale,
-                cameraData.isHdrEnabled, msaaSamples, needsAlphaChannel, cameraData.requiresOpaqueTexture);
+                cameraData.isHdrEnabled, asset.hdrColorBufferFormat, msaaSamples, needsAlphaChannel, cameraData.requiresOpaqueTexture);
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs:791]: Yes. Good catch.

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs (hunk 19 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipelineCore.cs:142]: I turned it into internal instead.

**Review by Jonasmortensen — COMMENTED** [review]
Works as intended. 
A couple of UI proposals:
Grey out HDR Format when HDR is not enabled
Rename "HDR Format" to "HDR Precision" and make the options "32 bits" and "64 bits" and update the tooltip

**Comment by TheoWong-pixel** [comment]
If users do not need to know about HDR Format when HDR is disabled, we should hide it from the users's view instead of greying it out. Otherwise i agree with Jonas' comments.

**Comment by eh-unity** [comment]
We haven't used exact format or bits elsewhere either. It's easier to change/update the formats if they're just abstract Low/High also we can't guarantee which formats a platform supports. I think, we should be either 100% explicit or abstract about the format. Currently abstract is the consistent way to go.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 14 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs]: per light now.

File: com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/PostProcessPasses.cs]: init to none to make clear it's just a struct init value.

**Comment by eh-unity** [comment]
Implemented the UI proposals. Attached some pics to the description.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 24]
```diff
@@ -113,6 +113,18 @@ public enum LightCookieFormat
...
     }
 
+    /// <summary>
+    /// Color Buffer format in HDR (only).
+    /// </summary>
+    public enum HDRColorBufferPrecision
+    {
+        /// <summary>Typically R11G11B10f for faster rendering. Recommend for mobile.
+        /// R11G11B10f can cause a subtle blue/yellow banding in some rare cases due to lower precision of blue component.</summary>
+        _32Bits,
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:123]: Maybe you can define an [InpsectorName] and a [Tooltip] attributes to nicify the inspector.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs [hunk 25]
```diff
@@ -836,6 +849,15 @@ public bool supportsHDR
...
         }
 
+        /// <summary>
+        /// Graphics format requested for HDR color buffers.
+        /// </summary>
+        public HDRColorBufferPrecision hdrColorBufferPrecision
+        {
+            get { return m_HDRColorBufferPrecision; }
+            set { m_HDRColorBufferPrecision = value; }
+        }
```
Comment (alex-vazquez) [inline com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:859]:
```suggestion
        public HDRColorBufferPrecision hdrColorBufferPrecision
        {
            get => m_HDRColorBufferPrecision;
            set => m_HDRColorBufferPrecision = value; 
        }
```

you can write this

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs (hunk 17 above)
Comment (alex-vazquez) [reply com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:26]: By default classes are internal, so maybe you only want to remove the public

**Review by alex-vazquez — APPROVED** [review]
Looks good, thanks, I've kicked the job for finding missing docs, just to know if you miss something.

**Review by ellioman — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 25 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:859]: While I agree it's nicer to read, all the other getters are in above style and it's the same amount of chars except for the return.
If we want to change this, I think we should change them all in a batch.

File: com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs (hunk 24 above)
Comment (eh-unity) [reply com.unity.render-pipelines.universal/Runtime/Data/UniversalRenderPipelineAsset.cs:125]: Added tooltips for the enum options.

**Review by Jonasmortensen — APPROVED** [review]
UI discussed

**Comment by eh-unity** [comment]
Test failures seem unrelated.
